docs: add pagination design pattern page (#1193)
pluralsight/pando: website/docs/learn/learn-pando/design-patterns/pagination.mdx

Reviewer: See Slack comment. We need to restructure to match the other pages. 👍

---
tags: [Pagination, lazy loading, large lists]
title: 'Displaying Large Lists'
---

We offer multiple pagination patterns to display large lists of content in more manageable chunks so you can optimize the experience based on the scenario.

## Manual Loading vs. Modern vs. Pagination Bar

### Manual Loading

Manual loading is like infinite scrolling, but the user controls when the next set of content is loaded. This approach appends the next set of content to the end of the list, allowing the user to continue scrolling with minimal disruption.

> This pattern works well in a "discovery" scenario when the user is browsing&mdash;not looking for anything specific&mdash;and aren't likely to need to revisit previously viewed content.

Reviewer: ```suggestion
This pattern works well in a "discovery" scenario when the user is browsing&mdash;not looking for anything specific&mdash;and **aren't likely to need to revisit previously viewed content**.

:::tip When to use this approach

Ask yourself: "If I was a user, would I care about needing an organized approach for viewing old and new content? 

Use this solution if you answered "no".

:::
```

### Modern

This approach displays only one set of content at a time while providing controls for the user to navigate forward or backward through the full list.

> This is useful when navigating ordered content, such as comments or course listings.

Reviewer: ```suggestion
This is useful when navigating ordered content, such as comments or course listings and is what [Github currently uses](https://github.com/pluralsight/pando/commits/main).

:::tip When to use this approach

Ask yourself: "If I was a user, would I care about needing a simple way to view older and newer content? 

Use this solution if you answered "yes".

:::
```

### Pagination Bar

> The Pagination Bar expands on the Modern aproach, offering more information and control. This approach adds controls for the page size (number of items displayed) and indicates the size of the list, along with the user's position within that list.

Reviewer: ```suggestion
The Pagination Bar expands on the Modern approach, offering more information and control to the user. This approach adds controls for the page size (number of items displayed) and indicates the size of the list, along with the user's position within that list.
```

> The extra controls can be helpful when searching for content within larger, more complex lists, such as data tables, and works well when paired with more granular search/filtering controls.

Reviewer: ```suggestion
The extra controls can be helpful when searching for content within larger, more complex lists, such as data tables, and works well when paired with more granular search/filtering controls.

:::tip when to use this approach

Ask yourself: "If I was a user, would I care about needing to view more information and have full control over how it is displayed and navigated through?
 
Use this solution if you answered "yes".

:::
```

## Modern for Simple Lists

For the Modern approach, we use two buttons that replace the displayed content with the next set from the indicated direction. If there is no more content in that direction, the corresponding button should be disabled.

```jsx live
function PaginationButtons() {
  const [activePage, setActivePage] = useState(0)
  const content = ['first page', 'second page', 'third page']
  const styles = getPaginationProps()
  const newerButtonProps = getButtonProps({
    ...styles.buttonOptions,
    disabled: activePage <= 0,
  })
  const olderButtonProps = getButtonProps({
    ...styles.buttonOptions,
    disabled: activePage >= content.length - 1,
  })

  function loadNewerContent() {
    setActivePage((prev) => prev - 1)
  }

  function loadOlderContent() {
    setActivePage((prev) => prev + 1)
  }

  return (
    <div>
      {content[activePage]}
      <div {...styles.buttonGroup}>
        <button {...newerButtonProps.button} onClick={loadNewerContent}>
          Newer
        </button>
        <button {...olderButtonProps.button} onClick={loadOlderContent}>
          Older
        </button>
      </div>
    </div>
  )
}
```

## Pagination Bar for Advanced Lists

> The pagination bar displays the total number of items in the list, as well as the items currently being displayed. A Select controls the number of items displayed at a time, and a pair of buttons are used to load the previous or next sets of data. If there is no data in that direction, the corresponding button should be disabled.

Reviewer: ```suggestion
For the Pagination Bar, we add page information combined with a Select component and the Modern approach appended to the end of the bar which shows disabled states based on the list data available for the user to navigate through.
```
